Modules/_datetimemodule.c

```diff
@@ -663,6 +663,144 @@ set_date_fields(PyDateTime_Date *self, int y, int m, int d)
     SET_DAY(self, d);
 }
 
+/* ---------------------------------------------------------------------------
+ * String parsing utilities and helper functions
+ */
+static inline unsigned int to_int(char ptr)
+{
+    return (unsigned int)(ptr - '0');
+}
+
+static inline int is_digit(unsigned int tmp)
+{
+    return tmp <= 9;
+}
+
+static inline const char* parse_digits(const char* ptr, int* var,
+                                       size_t num_digits)
+{
+    for (size_t i = 0; i < num_digits; ++i) {
+        int tmp = to_int(*(ptr++));
+        if (!is_digit(tmp)) { return NULL; }
```

Comment:
I would just use ANSI C `isdigit` on the character.

Reply:
I can actually just replace this with `tmp <= 9`, I think `isdigit` might be overkill. This `is_digit` relies on the specific way that `to_int` works - and I only used this particular combination because in early `Cython`-based tests my profiling indicated that `atoi` was a slow step.